src/gtc/passes/oir_optimizations/caches.py

```diff
@@ -0,0 +1,512 @@
+            lmin, lmax = fill_limits.get(field, (0, 0))
+            lmin = max(lmin, first_unfilled.get(field, lmin))
+            for offset in range(lmin, lmax + 1):
+                k_offset = common.CartesianOffset(
+                    i=0,
+                    j=0,
+                    k=offset if loop_order == common.LoopOrder.FORWARD else -offset,
+                )
+                fill_stmts.append(
+                    oir.AssignStmt(
+                        left=oir.FieldAccess(
+                            name=cache, dtype=symtable[field].dtype, offset=k_offset
+                        ),
+                        right=oir.FieldAccess(
+                            name=field, dtype=symtable[field].dtype, offset=k_offset
+                        ),
+                    )
+                )
+            first_unfilled[field] = lmax
+        return fill_stmts, first_unfilled
+
+    @classmethod
+    def _flush_stmts(
+        cls,
+        loop_order: common.LoopOrder,
+        section: oir.VerticalLoopSection,
+        flushing_fields: Dict[str, str],
+        symtable: Dict[str, Any],
+    ) -> List[oir.AssignStmt]:
+        """Generate flush statements for the given loop section.
+
+        Args:
+            loop_order: forward or backward order.
+            section: loop section to split.
+            flushing_fields: mapping from field names to cache names.
+
+        Returns:
+            A list of flush statements.
+        """
+        write_fields = AccessCollector.apply(section).write_fields()
+        flush_stmts = []
+        for field, cache in flushing_fields.items():
+            if field in write_fields:
+                flush_stmts.append(
+                    oir.AssignStmt(
+                        left=oir.FieldAccess(
+                            name=field,
+                            dtype=symtable[field].dtype,
+                            offset=common.CartesianOffset.zero(),
+                        ),
+                        right=oir.FieldAccess(
+                            name=cache,
+                            dtype=symtable[field].dtype,
+                            offset=common.CartesianOffset.zero(),
+                        ),
+                    )
+                )
+        return flush_stmts
+
+    def visit_VerticalLoop(
+        self,
+        node: oir.VerticalLoop,
+        *,
+        new_tmps: List[oir.Temporary],
+        symtable: Dict[str, Any],
+        new_symbol_name: Callable[[str], str],
+        **kwargs: Any,
+    ) -> oir.VerticalLoop:
+        filling_fields: Dict[str, str] = {
+            c.name: new_symbol_name(c.name)
+            for c in node.caches
+            if isinstance(c, oir.KCache) and c.fill
+        }
+        flushing_fields: Dict[str, str] = {
+            c.name: filling_fields[c.name] if c.name in filling_fields else new_symbol_name(c.name)
+            for c in node.caches
+            if isinstance(c, oir.KCache) and c.flush
+        }
```

Comment:
Not important, but I think a more explicit for-loop would be more readable here:

```suggestion
        filling_fields: Dict[str, str] = {}
        flushing_fields: Dict[str, str] = {}
        for c in node.caches:
            if isinstance(c, oir.KCache):
                if c.fill:
                    filling_fields[c.name] = new_symbol_name(c.name)
                if c.flush:
                    flushing_fields[c.name] = filling_fields.get(c.name, new_symbol_name(c.name))
    }
```